File: backend/Services/UserService.cs
using backend.Models;
using backend.Models.DTOs;
using backend.Repositories;

namespace backend.Services
{
    public class UserService : IUserService
    {
        private readonly IUserRepository _userRepository;

        public UserService(IUserRepository userRepository)
        {
            _userRepository = userRepository;
        }
        public async Task<User> GetProfileAsync(int userId)
        {
            var user = await _userRepository.GetByIdAsync(userId);
            if (user == null)
            {
                throw new KeyNotFoundException("User not found.");
            }
            return user;
        }

        public async Task ChangePasswordAsync(int userId, ChangePasswordRequestDto request)
        {
            var user = await _userRepository.GetByIdAsync(userId);
            if (user == null)
            {
                throw new KeyNotFoundException("Користувача не знайдено.");
            }

            if (!user.CheckPassword(request.CurrentPassword))
            {
                throw new InvalidOperationException("Неправильний поточний пароль.");
            }

            user.SetPassword(request.NewPassword);

            await _userRepository.UpdateAsync(user);
        }

        public async Task UpdateProfileAsync(int userId, UpdateProfileRequestDto request)
        {
            var user = await _userRepository.GetByIdAsync(userId);
            if (user == null)
            {
                throw new KeyNotFoundException("Користувача не знайдено.");
            }

            user.UpdateName(request.Name);
            user.UpdatePhone(request.Phone);

            await _userRepository.UpdateAsync(user);
        }
    }
}

File: backend/Services/userService.js
import userRepository from '../repositories/userRepository.js';

class UserService {
    async getCurrentUser(userId) {
        const user = await userRepository.findById(userId);
        if (!user) throw new Error('User not found');
        return user;
    }

    async updateUserProfile(userId, { name, phone, avatar }) {
        const updateData = { name, phone };
        if (avatar) updateData.avatar = avatar;

        const updatedUser = await userRepository.updateUserById(userId, updateData);
        if (!updatedUser) throw new Error("Failed to update user profile");
        return updatedUser;
    };

    async getUserLoyaltyData(userId) {
        const user = await userRepository.findUserById(userId);
        if (!user) throw new Error("User not found");

        const bonusHistory = await userRepository.getUserBonusTransactions(userId);

        return {
            bonusPoints: user.bonusPoints,
            bonusHistory,
        };
    };

}

export default new UserService();
